crates/vm/levm/src/precompiles.rs

```diff
@@ -678,12 +682,206 @@ fn update_pairing_result(
     Ok(())
 }
 
-fn blake2f(
-    _calldata: &Bytes,
-    _gas_for_call: u64,
-    _consumed_gas: &mut u64,
+// Message word schedule permutations for each round are defined by SIGMA constant.
+// Extracted from https://datatracker.ietf.org/doc/html/rfc7693#section-2.7
+pub const SIGMA: [[usize; 16]; 10] = [
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
+];
+
+// Initialization vector, used to initialize the work vector
+// Extracted from https://datatracker.ietf.org/doc/html/rfc7693#appendix-C.2
+pub const IV: [u64; 8] = [
+    0x6a09e667f3bcc908,
+    0xbb67ae8584caa73b,
+    0x3c6ef372fe94f82b,
+    0xa54ff53a5f1d36f1,
+    0x510e527fade682d1,
+    0x9b05688c2b3e6c1f,
+    0x1f83d9abfb41bd6b,
+    0x5be0cd19137e2179,
+];
+
+// Rotation constants, used in g
+// Extracted from https://datatracker.ietf.org/doc/html/rfc7693#section-2.1
+const R1: u32 = 32;
+const R2: u32 = 24;
+const R3: u32 = 16;
+const R4: u32 = 63;
+
+// The G primitive function mixes two input words, "x" and "y", into
+// four words indexed by "a", "b", "c", and "d" in the working vector
+// v[0..15].  The full modified vector is returned.
+// Based on https://datatracker.ietf.org/doc/html/rfc7693#section-3.1
+#[allow(clippy::indexing_slicing)]
+fn g(v: [u64; 16], a: usize, b: usize, c: usize, d: usize, x: u64, y: u64) -> [u64; 16] {
+    let mut ret = v;
+    ret[a] = v[a].wrapping_add(v[b]).wrapping_add(x);
+    ret[d] = (ret[d] ^ ret[a]).rotate_right(R1);
+    ret[c] = ret[c].wrapping_add(ret[d]);
+    ret[b] = (ret[b] ^ ret[c]).rotate_right(R2);
+    ret[a] = ret[a].wrapping_add(ret[b]).wrapping_add(y);
+    ret[d] = (ret[d] ^ ret[a]).rotate_right(R3);
+    ret[c] = ret[c].wrapping_add(ret[d]);
+    ret[b] = (ret[b] ^ ret[c]).rotate_right(R4);
```

Comment:
@maximopalopoli Is it possible that you are not applying the modulus operator here?

Reply:
See: https://datatracker.ietf.org/doc/html/rfc7693#section-3.1